hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieSnapshotExporter.java

```diff
@@ -0,0 +1,233 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.utilities;
+
+import com.beust.jcommander.JCommander;
+import com.beust.jcommander.Parameter;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.FileUtil;
+import org.apache.hadoop.fs.Path;
+import org.apache.hudi.common.util.StringUtils;
+import org.apache.hudi.common.SerializableConfiguration;
+import org.apache.hudi.common.model.HoodiePartitionMetadata;
+import org.apache.hudi.common.table.HoodieTableConfig;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.HoodieTimeline;
+import org.apache.hudi.common.table.TableFileSystemView;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.table.view.HoodieTableFileSystemView;
+import org.apache.hudi.common.util.FSUtils;
+import org.apache.hudi.common.util.Option;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.DataFrameWriter;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SaveMode;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.execution.datasources.DataSource;
+
+import scala.Tuple2;
+import scala.collection.JavaConversions;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Export the latest records of Hudi dataset to a set of external files (e.g., plain parquet files).
+ */
+
+public class HoodieSnapshotExporter {
+  private static final Logger LOG = LogManager.getLogger(HoodieSnapshotExporter.class);
+
+  public static class Config implements Serializable {
+    @Parameter(names = {"--source-base-path"}, description = "Base path for the source Hudi dataset to be snapshotted", required = true)
+    String sourceBasePath = null;
+
+    @Parameter(names = {"--target-base-path"}, description = "Base path for the target output files (snapshots)", required = true)
+    String targetOutputPath = null;
+
+    @Parameter(names = {"--snapshot-prefix"}, description = "Snapshot prefix or directory under the target base path in order to segregate different snapshots")
```

Comment:
the description here, I feel a little bit wierd after reading the code below, it means only export source Hudi dataset which contains the snapshotPrefix? would be changed to `Snapshot prefix or directory under the source Hudi dataset to be exported`? cc @OpenOpened @xushiyan

Reply:
@leesf thanks for catching this. I missed this one: this param is meant to segregate output not input. It was meant to be used in case for multiple exports with the same target path but wanted to be separated from each other (e.g., due to different export date). It actually overlaps with the target base path; users can simply change the target base paths to achieve this. So in conclusion we can just remove this param. @OpenOpened sounds good?

Reply:
I don't think we can delete this parameter. We rely on the metadata file .hoodie metadata in the root directory of the datasource to find things like `commitime`, valid parquet files, etc. If you point directly to the folder that needs to be exported, such as ROOT/2015/03/16 for the test case, an exception will be thrown `Hoodie table not found in path / tmp / junit4184871464195097137 / 2015/03/16 / .hoodie`.  I agree with @leesf, modify the comment of `--snapshot-perfix`.

Reply:
I see there're some gaps...this param `snapshot-prefix` is meant to let users export to a specific output directory. For example, --target-base-path=/mytable/ --snapshot-prefix=2020/03/03 and the output data will reside in `/mytable/2020/03/03/`. After removing --snapshot-prefix, the users will simply set --target-base-path=/mytable/2020/03/03/. This is a redundant param that should be removed from the RFC document too.

It is not meant to use on --source-base-path. Users will give the right base path to the hudi dataset to export like --source-base-path=/myhuditable/

Reply:
Talked with @xushiyan  offline, the RFC9 means we only export whole hudi dataset, thus may not support exporting `ROOT/2015/03/16`, thus we could remove `--snapshot-prefix` and modify the code accordingly. If we need export specified partitions later, we could bring it back and may need more consideration(maybe not use contains, maybe equals or regular expression). WDYT @OpenOpened